### PR TITLE
Add shape op and tests to TFLM

### PR DESCRIPTION
**Pull request by yair-ehrenwald**
Added SHAPE op and wrote a few tests.
SHAPE is required in order to run keras GRUCell layers in TFLM.

The necessity is further explained in this issue:
https://github.com/tensorflow/tensorflow/issues/43380

I left the switch/case for checking output types even though I removed support for int64 because I thought it might be convenient for adding types in the future (maybe even use int8?).

### CODE REVIEW

File: tensorflow/lite/micro/kernels/shape.cc [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  return kTfLiteOk;
+}
+
+TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+  const TfLiteEvalTensor* input =
+      tflite::micro::GetEvalInput(context, node, kInputTensor);
+  TfLiteEvalTensor* output =
+      tflite::micro::GetEvalOutput(context, node, kOutputTensor);
+
+  switch (output->type) {
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/shape.cc:54]: Not a huge deal one way or another, but I vote to remove this switch until we do add a second case.  Helps keep the code as clear and concise as possible.

**Review by njeffrie — DISMISSED** [review]
Quick style nit regarding the switch statement, otherwise good to go.

File: tensorflow/lite/micro/kernels/shape.cc (hunk 1 above)
Comment (yair-ehrenwald) [reply tensorflow/lite/micro/kernels/shape.cc]: Done :)

**Comment by yair-ehrenwald** [comment]
@advaitjain @njeffrie 
:)

File: tensorflow/lite/micro/kernels/shape.cc [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/kernels/op_macros.h"
+#include "tensorflow/lite/micro/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/memory_helpers.h"
+#include "tensorflow/lite/micro/micro_utils.h"
+
+namespace tflite {
+namespace ops {
+namespace micro {
+namespace shape {
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/shape.cc:28]: TFLM is incrementally moving towards a flat tflite namespace (https://abseil.io/tips/130)

can you remove this hierarchy and put the file-local functions into an anonymous namespace?

See [full_connected.cc](https://github.com/tensorflow/tensorflow/pull/43682/files#diff-c3c2bd6c6974d231aec4cd63d94fe47b) from this PR as an initial template.

File: tensorflow/lite/micro/kernels/shape.cc [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+constexpr int kInputTensor = 0;
+constexpr int kOutputTensor = 0;
+
+void ExtractShape(const TfLiteEvalTensor* input, int32_t* output_data) {
+  int numInputDims = input->dims->size;
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/shape.cc:34]:
```
const int num_input_dims = ...
```

or get rid of the temporary altogether.

File: tensorflow/lite/micro/kernels/shape.cc [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+  const TfLiteEvalTensor* input =
+      tflite::micro::GetEvalInput(context, node, kInputTensor);
+  TfLiteEvalTensor* output =
+      tflite::micro::GetEvalOutput(context, node, kOutputTensor);
+  if (output->type != kTfLiteInt32)
+    return kTfLiteError;
+  else
+    ExtractShape(input, tflite::micro::GetTensorData<int32_t>(output));
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/shape.cc:56]: nit: put braces around the single line if-else blocks too, for consistency.

File: tensorflow/lite/micro/kernels/shape.cc [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  return kTfLiteOk;
+}
+
+TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+  const TfLiteEvalTensor* input =
+      tflite::micro::GetEvalInput(context, node, kInputTensor);
+  TfLiteEvalTensor* output =
+      tflite::micro::GetEvalOutput(context, node, kOutputTensor);
+  if (output->type != kTfLiteInt32)
+    return kTfLiteError;
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/shape.cc:54]: add an error message?

```
TF_LITE_KERNEL_LOG(context, "Output type %s (%d) not supported.", TfLiteTypeGetName(output->type), output->type);
```

**Review by advaitjain — CHANGES_REQUESTED** [review]
only small style requests.

and the flatbuffer_conversions pull request is merged so let's merge master back into this PR.

File: tensorflow/lite/micro/kernels/shape.cc (hunk 2 above)
Comment (yair-ehrenwald) [reply tensorflow/lite/micro/kernels/shape.cc:28]: Where should I place the prototype for Register_SHAPE()? Is the idea for every kernel to have a header file that will be included in micro_mutable_op_resolver.h?

File: tensorflow/lite/micro/kernels/shape.cc (hunk 3 above)
Comment (yair-ehrenwald) [reply tensorflow/lite/micro/kernels/shape.cc]: done..

File: tensorflow/lite/micro/kernels/shape.cc (hunk 4 above)
Comment (yair-ehrenwald) [reply tensorflow/lite/micro/kernels/shape.cc]: done

File: tensorflow/lite/micro/kernels/shape.cc (hunk 5 above)
Comment (yair-ehrenwald) [reply tensorflow/lite/micro/kernels/shape.cc:56]: yes

File: tensorflow/lite/micro/kernels/shape.cc (hunk 2 above)
Comment (advaitjain) [reply tensorflow/lite/micro/kernels/shape.cc:29]: Good question. We'll move to a separate header on a case by csae basis, as needed for the code size optimization.

For this operator, let's keep Register_SHAPE if in  in micro_ops.h but move it out to the tflite namespace.

So micro_ops will become something like:

```cc

// Forward declaration of all micro op kernel registration methods. These
// registrations are included with the standard `BuiltinOpResolver`.
//
// This header is particularly useful in cases where only a subset of ops are
// needed. In such cases, the client can selectively add only the registrations
// their model requires, using a custom `(Micro)MutableOpResolver`. Selective
// registration in turn allows the linker to strip unused kernels.

namespace tflite {

// TFLM is incrementally moving towards a flat tflite namespace (https://abseil.io/tips/130). Any new ops (or cleanup of existing ops should have their Register function declarations in the tflite namespace.
TfLiteRegistration Register_SHAPE();


namespace ops {
namespace micro {
TfLiteRegistration Register_ABS();
TfLiteRegistration Register_ADD();
...

```

Comment (yair-ehrenwald) [reply tensorflow/lite/micro/kernels/shape.cc:29]: Ok. All done.

**Review by advaitjain — DISMISSED** [review]
I pushed some small changes that we don't have any existing examples for. Ready for merge.

**Review by advaitjain — APPROVED** [review]

**Pull request merged**